### PR TITLE
[Local GC] Move Software Write Watch's write barrier updates to GCToEEInterface::StompWriteBarrier

### PR DESCRIPTION
**Pull request by swgillespie**
This PR completes https://github.com/dotnet/coreclr/issues/8359 by removing the shared globals used by the write barrier and `softwarewritewatch.cpp`, which lies on the GC side of the GC-EE interface. The two shared globals are `g_sw_ww_table`, the table used by the write barrier for software write watch, and `g_sw_ww_enabled_for_gc_heap`, whether or not the GC has requested that write watch be active. In the same vein as what was done in https://github.com/dotnet/coreclr/pull/8568, the EE and GC share copies of these two global variables, which are synchronized in `GCToEEInterface::StompWriteBarrier`.  The GC calls `SoftwareWriteWatch::EnableForGCHeap` when it wants to track writes to the heap, and `SoftwareWriteWatch::DisableForGCHeap` when it is done - both calls end up calling `GCToEEInterface::StompWriteBarrier` and may bash the write barrier to a version that records writes to the heap using the software write watch table.

This PR also removes all inclusions of `softwarewritewatch.h` from within the EE.

### CODE REVIEW

**Comment by swgillespie** [comment]
@dotnet-bot test OSX Release longgc
@dotnet-bot test Ubuntu Release longgc

**Comment by swgillespie** [comment]
@dotnet-bot test Windows_NT Release standalone_gc

**Comment by swgillespie** [comment]
@dotnet-bot test Windows_NT Release standalone_gc
@dotnet-bot test OSX Release longgc
@dotnet-bot test Ubuntu Release longgc

cc @adityamandaleeka @jkotas @sergiy-k @Maoni0

File: src/vm/gcheaputilities.h [hunk 1]
```diff
@@ -108,22 +133,46 @@ class GCHeapUtilities {
...
+    // "dirty" (e.g. has been written to).
+    inline static void SoftwareWriteWatchSetDirty(void* address, size_t write_size)
+    {
+        LIMITED_METHOD_CONTRACT;
+
+        // We presumably have just written something to this address, so it can't be null.
+        assert(address != nullptr);
+
+        // The implementation is limited to writes of a pointer size or less. Writes of
+        // a pointer or larger may cross page boundaries and would require us to potentially
```
Comment (swgillespie) [inline src/vm/gcheaputilities.h]: you're right, that's what I meant to say - thanks.

**Comment by Maoni0** [comment]
>     uint8_t* ephemeral_lo;

I didn't get a chance to review the last PR on this issue so this is a comment on your last PR, can we use ephemeral_low and ephemeral_high like  we do in GC instead of ephemeral_lo and ephemeral_hi? 

---
Refers to: src/gc/gcinterface.h:90 in df94473. [](commit_id = df94473dbde77a6ae415656e3662b2aec470cf76, deletion_comment = False)

File: src/vm/gcheaputilities.h [hunk 2]
```diff
@@ -108,22 +133,46 @@ class GCHeapUtilities {
...
+        // set more than one entry in the SWW table, which can't be done atomically under
+        // the current scheme.
+        assert(write_size <= sizeof(void*));
+
+        size_t table_byte_index = reinterpret_cast<size_t>(address) >> SOFTWARE_WRITE_WATCH_AddressToTableByteIndexShift;
+        assert(table_byte_index != 0);
```
Comment (Maoni0) [inline src/vm/gcheaputilities.h]: >assert(table_byte_index != 0); [](start = 8, length = 30)

is there any reason to replace the original assert assert(GetTableByteIndex(reinterpret_cast<uint8_t *>(address) + (writeByteSize - 1)) == tableByteIndex); with this one?

**Comment by Maoni0** [comment]
@swgillespie ahh, right there's the macro. can we name them gc_ephemeral_low/high?

**Comment by swgillespie** [comment]
@Maoni0 will do! the macro, or the fields of `WriteBarrierParameters`?

File: src/vm/gcheaputilities.h (hunk 2 above)
Comment (swgillespie) [inline src/vm/gcheaputilities.h]: none in particular, I'll bring it over

**Comment by Maoni0** [comment]
@swgillespie  I looked at this a bit more - we really should get rid of #define ephemeral_* g_ephemeral_* to do a proper separation because you are already passing ephemeral_low/high as fields in the fields of WriteBarrierParameters. g_ephemeral_low/high are only used on the VM side - GC only uses its own, ie, ephemeral_low/high (it only looks at the g_* version in verify_heap). can you try to just get rid of the macros and see if you can compile? if you are getting a lot of errors you can just rename ephemeral_* to gc_ephemeral* in WriteBarrierParameters and the places these fields are used and be done with this PR. I can help you make the proper change after I get back.

File: src/gc/sample/GCSample.cpp [hunk 3]
```diff
@@ -94,7 +94,7 @@ inline void ErectWriteBarrier(Object ** dst, Object * ref)
     if (((uint8_t*)dst < g_gc_lowest_address) || ((uint8_t*)dst >= g_gc_highest_address))
         return;
         
-    if((uint8_t*)ref >= g_gc_ephemeral_low && (uint8_t*)ref < g_gc_ephemeral_high)
+    //if((uint8_t*)ref >= g_gc_ephemeral_low && (uint8_t*)ref < g_gc_ephemeral_high)
```
Comment (swgillespie) [inline src/gc/sample/GCSample.cpp]: I'll clean this up before check-in - removing g_gc_ephemeral_* broke the sample here.

**Comment by swgillespie** [comment]
@dotnet-bot test OSX Release longgc
@dotnet-bot test Ubuntu Release longgc

**Comment by swgillespie** [comment]
I just squashed all of my commits - this PR should be ready to merge as long as there aren't any objections.

**Comment by swgillespie** [comment]
Post-holiday ping - if there aren't any objections I'm hoping to get this merged soon and do a CoreRT GC refresh.

**Comment by Maoni0** [comment]
I'll look again this afternnon

File: src/gc/softwarewritewatch.h [hunk 4]
```diff
@@ -204,9 +199,13 @@ inline void SoftwareWriteWatch::EnableForGCHeap()
 
     VerifyCreated();
     assert(!IsEnabledForGCHeap());
+    g_gc_sw_ww_enabled_for_gc_heap = true;
 
-    g_sw_ww_enabled_for_gc_heap = true;
-    SwitchToWriteWatchBarrier(true);
+    WriteBarrierParameters args = {};
```
Comment (swgillespie) [inline src/gc/softwarewritewatch.h]: nice! will do.

File: src/gc/gc.cpp [hunk 5]
```diff
@@ -35739,7 +35745,13 @@ GCHeap::SetCardsAfterBulkCopy( Object **StartPoint, size_t len )
     end = StartPoint + (len/sizeof(Object*));
     while (rover < end)
     {
-        if ( (((uint8_t*)*rover) >= g_gc_ephemeral_low) && (((uint8_t*)*rover) < g_gc_ephemeral_high) )
+        // For Workstation GC, where there is exactly one ephemeral region, we can avoid touching the
+        // card table if the rover is currently pointing into a region outside of the ephemeral generation.
+        //
+        // For Server GC, where there are multiple ephemeral regions, we must always touch the card table.
+#ifndef MULTIPLE_HEAPS
+        if ( (((uint8_t*)*rover) >= gc_heap::ephemeral_low) && (((uint8_t*)*rover) < gc_heap::ephemeral_high) )
```
Comment (Maoni0) [inline src/gc/gc.cpp]: >if ( (((uint8_t*)*rover) >= gc_heap::ephemeral_low) && (((uint8_t*)*rover) < gc_heap::ephemeral_high) ) [](start = 8, length = 103)

this should not be changed - this function shouldn't know about gc_heap::ephemeral_low/high - it should use g_gc_* version as it's an external function to GC

**Comment by Maoni0** [comment]
>     ephemeral_high = heap_segment_reserved (ephemeral_heap_segment);

you should set g_gc_ephemeral_* here for workstation GC so external things can see the updated g_gc_ephemeral_*.

---
Refers to: src/gc/gc.cpp:9674 in 3952b64. [](commit_id = 3952b64de1c9ba02c5ae1db9eb8e8d970f77d175, deletion_comment = False)

File: src/gc/gc.cpp (hunk 5 above)
Comment (jkotas) [inline src/gc/gc.cpp]: Agree - it may be best to move this to the EE side, and remove it from the GC/EE interface. It is on the EE side in CoreRT already: https://github.com/dotnet/corert/search?utf8=%E2%9C%93&q=InlinedBulkWriteBarrier

**Comment by Maoni0** [comment]
> void gc_heap::adjust_ephemeral_limits (bool is_runtime_suspended)

can you please get rid of this arg? adjust_epheral_limits is only called when EE is suspended.

---
Refers to: src/gc/gc.cpp:9671 in 3952b64. [](commit_id = 3952b64de1c9ba02c5ae1db9eb8e8d970f77d175, deletion_comment = False)

**Comment by Maoni0** [comment]
>     stomp_write_barrier_ephemeral(is_runtime_suspended, ephemeral_low, ephemeral_high);

I would just call this only in workstation GC case and get rid of the code for server GC in it. actually it's better to just update the g_gc_ephemeral_* in there instead of in this method so it's clear that the ephemral range only matters in wks case.

---
Refers to: src/gc/gc.cpp:9680 in 3952b64. [](commit_id = 3952b64de1c9ba02c5ae1db9eb8e8d970f77d175, deletion_comment = False)

File: src/gc/gc.cpp (hunk 5 above)
Comment (swgillespie) [inline src/gc/gc.cpp]: ok, will do!

Comment (Maoni0) [inline src/gc/gc.cpp]: yep that would be best.

**Comment by Maoni0** [comment]
that was all my comments. sorry some of them didn't display correctly on GH...

File: src/gc/softwarewritewatch.h (hunk 4 above)
Comment (swgillespie) [inline src/gc/softwarewritewatch.h]: is this standard C++? MSVC rejects it and GCC issues a warning in pedantic mode that C99 style initializers like this aren't permitted in C++.

Comment (Maoni0) [inline src/gc/softwarewritewatch.h]: I would not bother with this.

Comment (janvorli) [inline src/gc/softwarewritewatch.h]: I am sorry, I did not know that this is really a C99 feature and not a C++ feature.  So please ignore my comment.

**Comment by swgillespie** [comment]
> I would just call this only in workstation GC case and get rid of the code for server GC in it. actually it's better to just update the g_gc_ephemeral_* in there instead of in this method so it's clear that the ephemral range only matters in wks case.

I ended up removing `g_gc_ephemeral_*` because they weren't used from within the GC, so I ran into trouble when doing this because, one way or another, the EE's `g_ephemeral_*` needs to be set somewhere, and `stomp_write_barrier_ephemeral` is the only place that does it. So, at least the way this PR is now, Server GC still needs to call `stomp_write_barrier_ephemeral` because it needs to give its `ephemeral_*` to the EE at least once, or else `ErectWriteBarrier` doesn't work.

I figure there are two solutions:

1. Remove calls to `stomp_write_barrier_ephemeral` under Server GC and initialize the EE's `g_ephemeral_low` and `g_ephemeral_high` to 1 and ~0 respectively in `stomp_write_barrier_initialize`. That way we don't need to call `stomp_write_barrier_ephemeral` at all in Server GC.
2. Remove calls to `stomp_write_barrier_ephemeral` under ServerGC and fix `ErectWriteBarrier` to not
check `g_ephemeral_low` and `g_ephemeral_high` when using Server GC, so that it unconditionally does the card table update. The EE already knows whether or not the GC in use is using Server GC or not so this check is not unreasonable.

Either one is fine with me, does anyone have any thoughts or opinions?

**Comment by jkotas** [comment]
> The EE already knows whether or not the GC in use is using Server GC

Ideally, we should be eliminating places where EE does explicitly things differently for Server GC. The GC should set g_ephemeral_low/g_ephemeral_high to 1 and ~0 (or they should be set to this value by default). The EE should look at the current values of  g_ephemeral_low/g_ephemeral_high, and if they are the boundary values, skip the ephemeral checks in the optimized write barrier.

**Comment by swgillespie** [comment]
Okay, will do!

**Comment by swgillespie** [comment]
Ubuntu leg hit #6574.

@dotnet-bot test Ubuntu x64 Checked

**Comment by swgillespie** [comment]
@dotnet-bot test Ubuntu x64 Checked Build and Test

**Comment by swgillespie** [comment]
thanks!
